File: ctoy/parser.c

#include <stdio.h>
#include <stdlib.h> // exit
#include <string.h> // strncpy

#include "scanner.h"
#include "parser.h"

#include "../extends/extensions.h"

typedef struct {
  Token current;
  Token previous;
	int had_error;
  char* error_message;
} Parser;

Parser parser;
ParserResult parser_result;

ByteCode* new_ByteCode(int capacity) {
  ByteCode* bytecode = _alloc(sizeof(ByteCode));
  bytecode->count = 0;
  bytecode->capacity = capacity == 0 ? 8 : capacity;
  bytecode->op = _alloc(sizeof(OpCode) * capacity);
  return bytecode;
}

void emit(OpCode op) {
  ByteCode* bc = parser_result.bytecode;
  if (bc->count == bc->capacity) {
    bc->capacity *= 2;
    bc->op = _realloc(bc->op, sizeof(OpCode) * bc->capacity);
  }
  bc->op[bc->count++] = op;
}

void print_bytecode() {
  ByteCode* bc = parser_result.bytecode;
  for (int i = 0; i < bc->count; i++) {
    // printf("%d\n", bc->op[i]);
    switch (bc->op[i]) {
      case OP_LITERAL:
        printf("OP_LITERAL %d\n", bc->op[++i]);
        break;
      case OP_ADD:
        printf("OP_ADD\n");
        break;
      case OP_SUBTRACT:
        printf("OP_SUBTRACT\n");
        break;
      case OP_MULTIPLY:
        printf("OP_MULTIPLY\n");
        break;
      case OP_DIVIDE:
        printf("OP_DIVIDE\n");
        break;
      case OP_NEGATE:
        printf("OP_NEGATE\n");
        break;
      default:
        printf("OP_UNKNOWN\n");
        break;
    }
  }
}

static int parse_int(const char* start, int length) {
  char* int_str = malloc(length + 1);
  strncpy(int_str, start, length);
  int_str[length] = '\0';
  int int_val = atoi(int_str);
  free(int_str);
  return int_val;
}

static void error(char* message) {
  if (parser.had_error) {
    return;
  }

  parser.error_message = message;
  parser.had_error = 1;
}

static void advance() {
  parser.previous = parser.current;
  parser.current = scan_token();
  if (parser.current.type == TOKEN_ERROR) {
    error("Error: unexpected character\n");
  }
}

static void consume(TokenType type, char* message) {
  if (parser.current.type == type) {
    advance();
    return;
  }

  error(message);
}

static int get_precedence(TokenType type) {
  switch (type) {
    case TOKEN_PLUS:
    case TOKEN_MINUS:
      return 50;
    case TOKEN_STAR:
    case TOKEN_SLASH:
      return 60;
    default:
      return -1;
  }
}

// forward declaration
static void parse_expression(int precedence);

static void prefix_handler() {
  int int_val;
  int int_idx;

	switch (parser.previous.type) {
    case TOKEN_INT:
      int_val = parse_int(parser.previous.start, parser.previous.length);
      int_idx = append(parser_result.literals, new_Int(int_val));
      // -> emit byte code
      // INT idx
      emit(OP_LITERAL);
      emit(int_idx);
      
      printf("INT: %.*s\n", parser.previous.length, parser.previous.start);
      break;
    case TOKEN_MINUS:
			parse_expression(70);
      emit(OP_NEGATE);
      printf("SIGN MINUS\n");
      break;
    case TOKEN_LEFT_PAREN:
      parse_expression(0);
      consume(TOKEN_RIGHT_PAREN, "Error: expected right paren\n");
      break;
    default:
      error("Error: expected prefix token\n");
  }
}

static void infix_handler() {
	// save the token here for later
	TokenType type = parser.previous.type;

	// recurse
  parse_expression(get_precedence(type));

  switch (type) {
    case TOKEN_PLUS:
      printf("PLUS\n");
      emit(OP_ADD);
      break;
    case TOKEN_MINUS:
      printf("MINUS\n");
      emit(OP_SUBTRACT);
      break;
    case TOKEN_STAR:
      printf("STAR\n");
      emit(OP_MULTIPLY);
      break;
    case TOKEN_SLASH:
      printf("SLASH\n");
      emit(OP_DIVIDE);
      break;
    default:
      error("Error: expected infix token\n");
  }
}

static void parse_expression(int precedence) {

	advance();
	prefix_handler();
	// if (parser.had_error) {
 //    return;
 //  }

	while (precedence <= get_precedence(parser.current.type)) {
		advance();
		infix_handler();
		// if (parser.had_error) {
  //     return;
  //   }
  }

}

void interpret(char *source) {
  init_scanner(source);

	parser.had_error = 0;
  parser_result.literals = new_Array(0);
  parser_result.bytecode = new_ByteCode(0);

	advance();
  if (parser.current.type == TOKEN_EOF) {
    return;
  }

	parse_expression(0);
	consume(TOKEN_EOF, "Error: expected end of expression\n");

  if (parser.had_error) {
    printf("%s", parser.error_message);
  }
  // print size
  printf("sizeof: %lu\n", sizeof(long long int));
  
  printf("ByteCode:\n");
  print_bytecode();
  printf("Literals: ");
  print(parser_result.literals);
  printf("\n");
	/*
	while (1) {
    Token token = scanToken();

    if (token.type == TOKEN_EOF) break;

    printf("token type: %d\n", token.type);
		printf("token: \"%.*s\"\n", token.length, token.start);

  }
	*/

}

File: rtoy/src/parser.rs
// use crate::scanner::Scanner;
use crate::scanner::Token;
use crate::scanner::TokenType;

use crate::element::Element;

#[derive(Debug, Clone)]
pub enum Instruction {
    Add,
    Sub,
    Mul,
    Div,
    Neg,
    Lit(usize),
}

pub struct ParserResult {
    pub instructions: Vec<Instruction>,
    pub literals: Vec<Element>,
    pub had_error: bool,
    pub error_message: String,
}

pub struct Parser<'a> {
    current_token_idx: usize,
    source: &'a str,
    tokens: Vec<Token>,
    result: ParserResult,
}

impl<'a> Parser<'a> {
    pub fn parse(tokens: Vec<Token>, source: &str) -> ParserResult {
        let mut parser = Parser {
            current_token_idx: 0,
            source,
            tokens,
            result: ParserResult {
                instructions: Vec::new(),
                literals: Vec::new(),
                had_error: false,
                error_message: String::new(),
            },
        };
        parser.parse_expression(0);
        parser.result
    }

    fn error(&mut self, message: &str) {
        if self.result.had_error {
            return;
        }
        self.result.had_error = true;
        self.result.error_message = message.to_string();
    }

    fn consume(&mut self, token_type: TokenType) {
        if self.get_current_token().token_type == token_type {
            self.current_token_idx += 1;
            return;
        }
        self.error("Error: Expected token not found");
    }

    fn get_current_token(&self) -> &Token {
        &self.tokens[self.current_token_idx]
    }

    fn get_previous_token(&self) -> &Token {
        &self.tokens[self.current_token_idx - 1]
    }

    fn get_precedence(&self, token: &Token) -> i8 {
        match token.token_type {
            TokenType::Plus | TokenType::Minus => 50,
            TokenType::Star | TokenType::Slash => 60,
            _ => -1,
        }
    }

    fn get_token_value(&self, token: &Token) -> usize {
        self.source[token.start..token.end]
            .parse::<usize>()
            .expect("Failed to parse token value")
    }

    fn prefix_handler(&mut self) {
        let token = self.get_previous_token();
        match token.token_type {
            TokenType::Literal => {
                let literal = self.get_token_value(token);

                self.result.literals.push(Element::Int(literal as isize));

                self.result
                    .instructions
                    .push(Instruction::Lit(self.result.literals.len() - 1));
            }
            TokenType::Minus => {
                self.parse_expression(70);
                self.result.instructions.push(Instruction::Neg);
            }
            TokenType::Left_Paren => {
                self.parse_expression(0);
                self.consume(TokenType::Right_Paren);
            }
            _ => {
                self.error("Error: Unexpected token");
            }
        }
    }

    fn infix_handler(&mut self) {
        let token = self.get_previous_token().clone();

        self.parse_expression(self.get_precedence(&token));

        match token.token_type {
            TokenType::Plus => {
                self.result.instructions.push(Instruction::Add);
            }
            TokenType::Minus => {
                self.result.instructions.push(Instruction::Sub);
            }
            TokenType::Star => {
                self.result.instructions.push(Instruction::Mul);
            }
            TokenType::Slash => {
                self.result.instructions.push(Instruction::Div);
            }
            _ => {
                self.error("Error: Unexpected token");
            }
        }
    }

    fn parse_expression(&mut self, rbp: i8) {
        self.current_token_idx += 1;
        self.prefix_handler();

        loop {
            if rbp > self.get_precedence(&self.get_current_token()) {
                return;
            }
            self.current_token_idx += 1;

            self.infix_handler();
        }
    }
}
